fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/MultiJoinNode.java

```diff
@@ -112,19 +112,20 @@ private static void flattenJoinNode(OptExpression node, LinkedHashSet<OptExpress
 
         if (joinOperator.getProjection() != null) {
             Projection projection = joinOperator.getProjection();
-
-            for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projection.getColumnRefMap().entrySet()) {
-                if (!entry.getValue().isColumnRef()
-                        && entry.getValue().getUsedColumns().isIntersect(node.inputAt(0).getOutputColumns())
-                        && entry.getValue().getUsedColumns().isIntersect(node.inputAt(1).getOutputColumns())) {
-                    atoms.add(node);
-                    return;
-                }
-
-                if (!entry.getKey().equals(entry.getValue())) {
-                    expressionMap.put(entry.getKey(), entry.getValue());
-                }
+            boolean fromTwoChildren = projection.getColumnRefMap().values().stream().anyMatch(v -> {
+                ColumnRefSet useRefs = v.getUsedColumns();
+                return !v.isColumnRef() && useRefs.isIntersect(node.inputAt(0).getOutputColumns())
+                        && useRefs.isIntersect(node.inputAt(1).getOutputColumns());
+            });
+            if (fromTwoChildren) {
+                atoms.add(node);
+                return;
             }
+            projection.getColumnRefMap().forEach((k, v) -> {
```

Comment:
add into projection only when the operator doesn't belong to each child and not column ref?

Reply:
add into projection when the operator only belong to **one** child, meanings the expression can be execute in child node before the join